provider/google : add a new resource : google_compute_snapshot

tpoindessous opened the pull request:
Hi

here is a WIP for addressing issue https://github.com/hashicorp/terraform/issues/11690 

This PR adds a new resource : google_compute_snapshot

For today, I only tested creating and deleting a simple snapshot.

Also, I don't know how the documentation is made.

I will add tests acceptance in next days.

Thanks.

tpoindessous commented:
Hi,

my test acceptance doesn't work for the moment, terraform doesn't handle very the delete command, so created disks are not deleted yet.

Thanks

tpoindessous commented:
Hi

I corrected the test acceptance. Everything works now.

Thanks.

tpoindessous commented:
Hi @paddyforan 

I think this PR is ready to be reviewed, I added the documentation. Please tell me if I need to change something.

Thanks !

paddyforan commented:
Hey @tpoindessous! Thanks for the PR! I literally have a browser tab open to review it, but unfortunately, it's 2:30 in the morning here, so I need to get some sleep. I'm not sure I'll get to it tomorrow, as I already queued myself a pretty big workload (trying to recover from Next!) but this is on my radar and will get a review ASAP. I just need to set aside some time to do a deep dive and understand it better, along with the problem space. Thanks so much for the patience and PR, however; both are incredibly appreciated.

paddyforan commented on builtin/providers/google/resource_compute_snapshot.go:
```diff
@@ -0,0 +1,184 @@
+				Computed: true,
+			},
+
+			"sourcedisk_encryption_key_raw": &schema.Schema{
```
`sourcedisk` => `source_disk`?

paddyforan commented on builtin/providers/google/resource_compute_snapshot.go:
```diff
@@ -0,0 +1,184 @@
+				Sensitive: true,
+			},
+
+			"sourcedisk_encryption_key_sha256": &schema.Schema{
```
`sourcedisk` => `source_disk`?

paddyforan commented on builtin/providers/google/resource_compute_snapshot.go:
```diff
@@ -0,0 +1,184 @@
+				Computed: true,
+			},
+
+			"sourcedisk_id": &schema.Schema{
```
`sourcedisk` => `source_disk`?

paddyforan commented on builtin/providers/google/resource_compute_snapshot.go:
```diff
@@ -0,0 +1,184 @@
+				Computed: true,
+			},
+
+			"sourcedisk": &schema.Schema{
```
`sourcedisk` => `source_disk`?

paddyforan commented on builtin/providers/google/resource_compute_snapshot.go:
```diff
@@ -0,0 +1,184 @@
+	d.Set("self_link", snapshot.SelfLink)
+	if snapshot.SnapshotEncryptionKey != nil && snapshot.SnapshotEncryptionKey.Sha256 != "" {
+		d.Set("snapshot_encryption_key_sha256", snapshot.SnapshotEncryptionKey.Sha256)
+	}
```
It may sounds silly, but do you mind setting the other fields here to what the API returns? Especially `sourcedisk_encryption_key_sha256`, but the others too. I know they're in the config, but setting what the API returns will help Terraform detect if something on the API side gets out of sync with something on the Terraform side. For this resource it isn't as big a deal, but it's better to follow the best practice, unless we can't.

tpoindessous replied:
Hi, just to be sure : I need to duplicate lines 149-151 for each other fields returned by the API ?
Thanks

tpoindessous replied:
Forget my question. I just understood that my code was missing setting computed fields.

Thanks

paddyforan replied:
I'm not sure the `if` is necessary for every one, but yeah, ideally, we'd have a `Set` for every field in the resource that the API returns.

This is because the `Set` is how data from the API gets into the state, and Terraform diffs the state and the config to find if there's a difference. So the only way Terraform knows if the API and the config don't match is if we `Set` the data appropriately.

paddyforan commented on builtin/providers/google/resource_compute_snapshot.go:
```diff
@@ -0,0 +1,184 @@
+	return &schema.Resource{
+		Create: resourceComputeSnapshotCreate,
+		Read:   resourceComputeSnapshotRead,
+		Delete: resourceComputeSnapshotDelete,
```
I'd love to see an `Exists` function, as well!

paddyforan commented on builtin/providers/google/resource_compute_snapshot_test.go:
```diff
@@ -0,0 +1,158 @@
+
+		_, err := config.clientCompute.Snapshots.Get(
+			config.Project, rs.Primary.ID).Do()
+		if err == nil {
```
What happens if there's a server error/network error/anything but a 404?

tpoindessous replied:
Would be it OK to add this test :

`if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == 404 {`

paddyforan replied:
That would be perfect!

paddyforan replied:
(Assuming, of course, we check for non-404 errors and return that error, as well.)

paddyforan commented on builtin/providers/google/resource_compute_snapshot_test.go:
```diff
@@ -0,0 +1,158 @@
+	}
+}
+
+func testAccCheckSnapshotEncryptionKey(n string, snapshot *compute.Snapshot) resource.TestCheckFunc {
```
I'd love to see checks for the other fields, as well. They could probably be combined into testAccCheckComputeSnapshotExists.

paddyforan commented on website/source/docs/providers/google/r/compute_snapshot.html.markdown:
```diff
@@ -0,0 +1,70 @@
+
+* `zone` - (Required) The zone where the source disk is located.
+
+* `disk` - (Required)
```
A description would be great here.

paddyforan commented on website/source/docs/providers/google/r/compute_snapshot.html.markdown:
```diff
@@ -0,0 +1,70 @@
+
+- - -
+
+* `sourcedisk_encryption_key_raw` - (Optional) A 256-bit [customer-supplied encryption key]
```
Think this needs to be updated to `source_disk_encryption_key_raw`.

paddyforan commented on website/source/docs/providers/google/r/compute_snapshot.html.markdown:
```diff
@@ -0,0 +1,70 @@
+    [customer-supplied encryption key](https://cloud.google.com/compute/docs/disks/customer-supplied-encryption)
+    that protects this resource.
+
+* `sourcedisk_encryption_key_sha256` - The [RFC 4648 base64]
```
Think this needs to be updated to `source_disk_encryption_key_sha256`.

paddyforan commented on website/source/docs/providers/google/r/compute_snapshot.html.markdown:
```diff
@@ -0,0 +1,70 @@
+    [customer-supplied encryption key](https://cloud.google.com/compute/docs/disks/customer-supplied-encryption)
+    that protects the source disk.
+
+* `sourcedisk_id` - The ID value of the source disk used to create this snapshot.
```
Think this needs to be updated to `source_disk_id`.

paddyforan replied:
Also, how is this different from `source_disk`?

paddyforan commented on website/source/docs/providers/google/r/compute_snapshot.html.markdown:
```diff
@@ -0,0 +1,70 @@
+    This value may be used to determine whether the snapshot was taken from the
+    current or a previous instance of a given disk name.
+
+* `sourcedisk` - The source disk used to create this snapshot.
```
Think this needs to be updated to `source_disk`.

paddyforan commented on website/source/docs/providers/google/r/compute_snapshot.html.markdown:
```diff
@@ -0,0 +1,70 @@
+
+* `zone` - (Required) The zone where the source disk is located.
+
+* `disk` - (Required) The disk which will be used as the source of the snapshot
```
How is `disk` different from `source_disk`? Won't they always match?

tpoindessous commented:
Hi @paddyforan 

I first tried to replicate Google API to Terraform, but since Terraform keeps the state, some attributes are not needed, as you requested.

I dropped sourcedisk_id and disk attributes.

Thanks.

tpoindessous commented:
Hi @paddyforan 

sorry to bother you ... I take the risk to ping you about this PR :)

Thanks for your time :)

paddyforan commented:
Hey @tpoindessous! No risk in pinging--it is my job to review it, after all! You were generous enough to write code to add the feature, the least I can do is look at it. Apologies for the delay on this, I could've _sworn_ I replied to it already. I'll take a look right now and hopefully we can get this in the 0.9.4 release; you've certainly been patient enough about it.

paddyforan commented on builtin/providers/google/resource_compute_snapshot.go:
```diff
@@ -0,0 +1,205 @@
+		d.Set("source_disk_encryption_key_sha256", snapshot.SourceDiskEncryptionKey.Sha256)
+	}
+
+	d.Set("source_disk", snapshot.SourceDisk)
```
I'd love to see

```go
d.Set("name", snapshot.Name)
```

Here as well, but I'm not willing to hold this any longer over it.

paddyforan commented on website/source/docs/providers/google/r/compute_snapshot.html.markdown:
```diff
@@ -0,0 +1,64 @@
+---
```
We're going to need to add this resource to the menu as well, but I can always fix that up after.

paddyforan commented on builtin/providers/google/resource_compute_snapshot_test.go:
```diff
@@ -0,0 +1,219 @@
+
+resource "google_compute_snapshot" "foobar" {
+	name = "%s"
+	disk = "${google_compute_disk.foobar.name}"
```
Tests don't run for me, looks like you want `source_disk` here.

paddyforan commented on builtin/providers/google/resource_compute_snapshot_test.go:
```diff
@@ -0,0 +1,219 @@
+}
+resource "google_compute_snapshot" "foobar" {
+	name = "%s"
+	disk = "${google_compute_disk.foobar.name}"
```
Tests don't run for me, looks like you want source_disk here.

tpoindessous commented:
Thanks @paddyforan 

Sorry for the failing tests ...

I will do all your points.

I'm stuck with a design error : gcp.sourceDisk is an URL when TF source_disk is a name. So it's not matching. I will look at it in the next days, maybe I will revert change to use "disk" in TF and exposing source_disk only in output (it can be useful). But I need to find the right word in documentation.

I will come back to you only when tests will be OK and all your points are corrected.

Have a nice week-end and a nice Easter.

tpoindessous commented:
Hi

I corrected all the remarks you made. I added the new resource doc in the nav panel.

the two tests are OK : 
$ make testacc TEST=./builtin/providers/google TESTARGS='-run=TestAccComputeSnapshot_basic'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/04/20 12:51:32 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/google -v -run=TestAccComputeSnapshot_basic -timeout 120m
=== RUN   TestAccComputeSnapshot_basic
--- PASS: TestAccComputeSnapshot_basic (72.11s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/google	72.130s

$ make testacc TEST=./builtin/providers/google TESTARGS='-run=TestAccComputeSnapshot_encryption'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/04/20 12:53:26 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/google -v -run=TestAccComputeSnapshot_encryption -timeout 120m
=== RUN   TestAccComputeSnapshot_encryption
--- PASS: TestAccComputeSnapshot_encryption (72.97s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/google	72.986s

Thanks !

paddycarver commented:
Hey @tpoindessous!

> Sorry for the failing tests ...

No worries! It happens to the best of us. :)

> I'm stuck with a design error : gcp.sourceDisk is an URL when TF source_disk is a name. So it's not matching. I will look at it in the next days, maybe I will revert change to use "disk" in TF and exposing source_disk only in output (it can be useful). But I need to find the right word in documentation.

Last question on this:

Is there a reason not to accept the name _or_ URL for the `source_disk` parameter, and just expand to the URL? That would fix the problem, right?

Thanks so much for your effort and patience on this. If this is getting too drawn-out and your interest wanes, just let me know and I'm happy to do the work to push it over the finish line. :) You've been amazing, but I don't want to assume your generosity and patience are infinite.

(Also, I _promise_ a quick review and reply.)

nodesocket commented:
👍  on this PR. I wonder if it makes sense once this is merged and deployed to take backup snapshots of disks using Terraform? Then the question becomes, how to schedule them.

tpoindessous commented:
Hi @paddycarver 

we can't use get method of Disk object with a URL. Maybe we could use this method
https://cloud.google.com/compute/docs/reference/latest/disks/aggregatedList to find the right Disk but I think this may be a lot of work ... I'm not sure I can do it.

Can you please review my last commit ? I added some tests to be sure that the source_disk matches the same disk as source_disk_link.

I have prepared a new PR to add dataSourceGoogleComputeSnapshot, I think I will push it in the next 2 weeks.

Thanks.

Thomas

paddycarver commented:
> we can't use get method of Disk object with a URL.

I guess I'm confused. The URL contains the Disk name, couldn't we just pull it out of that? I'd like to, if possible, accept either here and just Do The Right Thing. But I also feel like I'm missing context. I'm going to read through all the code again right now and see if I can spot the problem.
